In [157]:
import pandas as pd
import numpy as np
from scipy import stats

### Phase 1 — Load & clean (Pandas)

Load all 3 files with correct separators and column names <br>

In [19]:
movies_df = pd.read_csv('~/Downloads/ml-32m/movies.csv')
movies_df.head()

,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy


In [16]:
ratings_df = pd.read_csv('~/Downloads/ml-32m/ratings.csv')
ratings_df.head()

,userId,movieId,rating,timestamp
0,1,17,4.0,944249077
1,1,25,1.0,944250228
2,1,29,2.0,943230976
3,1,30,5.0,944249077
4,1,32,5.0,943228858


In [17]:
tags_df = pd.read_csv('~/Downloads/ml-32m/tags.csv')
tags_df.head()

,userId,movieId,tag,timestamp
0,22,26479,Kevin Kline,1583038886
1,22,79592,misogyny,1581476297
2,22,247150,acrophobia,1622483469
3,34,2174,music,1249808064
4,34,2174,weird,1249808102


In [18]:
links_df = pd.read_csv('~/Downloads/ml-32m/links.csv')
links_df.head()

,movieId,imdbId,tmdbId
0,1,114709,862.0
1,2,113497,8844.0
2,3,113228,15602.0
3,4,114885,31357.0
4,5,113041,11862.0


### Handle dtypes, parse timestamps, identify and deal with missings and duplicates 

In [24]:
movies_df.head()

,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy


In [27]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87585 entries, 0 to 87584
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  87585 non-null  int64 
 1   title    87585 non-null  object
 2   genres   87585 non-null  object
dtypes: int64(1), object(2)
memory usage: 2.0+ MB


In [38]:
movies_df.isnull().sum()

movieId    0
title      0
genres     0
dtype: int64

In [28]:
movies_df.describe()

,movieId
count,87585.000000
mean,157651.365519
std,79013.402099
min,1.000000
25%,112657.000000
50%,165741.000000
75%,213203.000000
max,292757.000000


In [29]:
ratings_df.head()

,userId,movieId,rating,timestamp
0,1,17,4.0,944249077
1,1,25,1.0,944250228
2,1,29,2.0,943230976
3,1,30,5.0,944249077
4,1,32,5.0,943228858


In [30]:
ratings_df.describe()

,userId,movieId,rating,timestamp
count,3.200020e+07,3.200020e+07,3.200020e+07,3.200020e+07
mean,1.002785e+05,2.931861e+04,3.540396e+00,1.275241e+09
std,5.794905e+04,5.095816e+04,1.058986e+00,2.561630e+08
min,1.000000e+00,1.000000e+00,5.000000e-01,7.896520e+08
25%,5.005300e+04,1.233000e+03,3.000000e+00,1.051012e+09
50%,1.002970e+05,3.452000e+03,3.500000e+00,1.272622e+09
75%,1.504510e+05,4.419900e+04,4.000000e+00,1.503158e+09
max,2.009480e+05,2.927570e+05,5.000000e+00,1.697164e+09


In [31]:
ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32000204 entries, 0 to 32000203
Data columns (total 4 columns):
 #   Column     Dtype  
---  ------     -----  
 0   userId     int64  
 1   movieId    int64  
 2   rating     float64
 3   timestamp  int64  
dtypes: float64(1), int64(3)
memory usage: 976.6 MB


In [32]:
ratings_df['time_dt'] = pd.to_datetime(ratings_df['timestamp'])

In [33]:
ratings_df.head()

,userId,movieId,rating,timestamp,time_dt
0,1,17,4.0,944249077,1970-01-01 00:00:00.944249077
1,1,25,1.0,944250228,1970-01-01 00:00:00.944250228
2,1,29,2.0,943230976,1970-01-01 00:00:00.943230976
3,1,30,5.0,944249077,1970-01-01 00:00:00.944249077
4,1,32,5.0,943228858,1970-01-01 00:00:00.943228858


In [37]:
#check for nulls
ratings_df.isnull().sum()

userId       0
movieId      0
rating       0
timestamp    0
time_dt      0
dtype: int64

In [39]:
tags_df.head()

,userId,movieId,tag,timestamp
0,22,26479,Kevin Kline,1583038886
1,22,79592,misogyny,1581476297
2,22,247150,acrophobia,1622483469
3,34,2174,music,1249808064
4,34,2174,weird,1249808102


In [40]:
tags_df['time_dt']=pd.to_datetime(tags_df['timestamp'])

In [41]:
tags_df.head()

,userId,movieId,tag,timestamp,time_dt
0,22,26479,Kevin Kline,1583038886,1970-01-01 00:00:01.583038886
1,22,79592,misogyny,1581476297,1970-01-01 00:00:01.581476297
2,22,247150,acrophobia,1622483469,1970-01-01 00:00:01.622483469
3,34,2174,music,1249808064,1970-01-01 00:00:01.249808064
4,34,2174,weird,1249808102,1970-01-01 00:00:01.249808102


In [45]:
#check for nulls
tags_df.isnull().sum()

userId        0
movieId       0
tag          17
timestamp     0
time_dt       0
dtype: int64

In [43]:
tags_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000072 entries, 0 to 2000071
Data columns (total 5 columns):
 #   Column     Dtype         
---  ------     -----         
 0   userId     int64         
 1   movieId    int64         
 2   tag        object        
 3   timestamp  int64         
 4   time_dt    datetime64[ns]
dtypes: datetime64[ns](1), int64(3), object(1)
memory usage: 76.3+ MB


In [44]:
tags_df.describe()

,userId,movieId,timestamp,time_dt
count,2.000072e+06,2.000072e+06,2.000072e+06,2000072
mean,8.192859e+04,7.189326e+04,1.528914e+09,1970-01-01 00:00:01.528913885
min,2.200000e+01,1.000000e+00,1.135429e+09,1970-01-01 00:00:01.135429210
25%,6.841300e+04,4.011000e+03,1.473615e+09,1970-01-01 00:00:01.473614704
50%,7.821300e+04,5.232800e+04,1.574071e+09,1970-01-01 00:00:01.574071062
75%,1.036980e+05,1.222940e+05,1.614740e+09,1970-01-01 00:00:01.614739619
max,1.622790e+05,2.926290e+05,1.697155e+09,1970-01-01 00:00:01.697154983
std,3.810650e+04,7.480379e+04,1.290835e+08,NaN


In [47]:
tags_df.count()

userId       2000072
movieId      2000072
tag          2000055
timestamp    2000072
time_dt      2000072
dtype: int64

<b>The dataset is nearly complete for the tag field, with only 17 missing values out of ~2 million rows, so missing data handling will have minimal impact.

In [49]:
#get null percentage
tags_df.isnull().mean()*100

userId       0.00000
movieId      0.00000
tag          0.00085
timestamp    0.00000
time_dt      0.00000
dtype: float64

In [51]:
#fill missing values in Tags columns with 'Unknown'

tags_df['tag']=tags_df['tag'].fillna('Unknown')

In [55]:
#recheck nulls- there should be no nulls
tags_df.isnull().mean()

userId       0.0
movieId      0.0
tag          0.0
timestamp    0.0
time_dt      0.0
dtype: float64

In [57]:
#check duplicates

dupes= ratings_df.duplicated(subset=['userId','movieId'])

dupes.sum()

np.int64(0)

In [64]:
#extract year from movie title
movies_df['year'] = movies_df['title'].str.extract(r'\((\d{4})\)')

movies_df['year']=pd.to_numeric(movies_df['year'],errors='coerce')

movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87585 entries, 0 to 87584
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   movieId  87585 non-null  int64  
 1   title    87585 non-null  object 
 2   genres   87585 non-null  object 
 3   year     86970 non-null  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 2.7+ MB


In [65]:
 v

,movieId,title,genres,year
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,1995.0
1,2,Jumanji (1995),Adventure|Children|Fantasy,1995.0
2,3,Grumpier Old Men (1995),Comedy|Romance,1995.0
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance,1995.0
4,5,Father of the Bride Part II (1995),Comedy,1995.0


In [66]:
genre_dummies= movies_df['genres'].str.get_dummies('|') 
genre_dummies

,(no genres listed),Action,Adventure,Animation,Children,Comedy,Crime,Documentary,Drama,Fantasy,Film-Noir,Horror,IMAX,Musical,Mystery,Romance,Sci-Fi,Thriller,War,Western
0,0,0,1,1,1,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0
1,0,0,1,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0
2,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0
3,0,0,0,0,0,1,0,0,1,0,0,0,0,0,0,1,0,0,0,0
4,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
87580,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0
87581,0,0,0,0,0,1,0,0,1,0,0,0,0,0,0,0,0,0,0,0
87582,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0
87583,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0


In [75]:
movies_df=pd.concat([movies_df,genre_dummies],axis=1)
movies_df.head()

,movieId,title,genres,year,(no genres listed),Action,Adventure,Animation,Children,Comedy,...,Film-Noir,Horror,IMAX,Musical,Mystery,Romance,Sci-Fi,Thriller,War,Western
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,1995.0,0,0,1,1,1,1,...,0,0,0,0,0,0,0,0,0,0
1,2,Jumanji (1995),Adventure|Children|Fantasy,1995.0,0,0,1,0,1,0,...,0,0,0,0,0,0,0,0,0,0
2,3,Grumpier Old Men (1995),Comedy|Romance,1995.0,0,0,0,0,0,1,...,0,0,0,0,0,1,0,0,0,0
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance,1995.0,0,0,0,0,0,1,...,0,0,0,0,0,1,0,0,0,0
4,5,Father of the Bride Part II (1995),Comedy,1995.0,0,0,0,0,0,1,...,0,0,0,0,0,0,0,0,0,0


In [74]:
ratings_df.head()

,userId,movieId,rating,timestamp,time_dt,rating_year,rating_month
0,1,17,4.0,944249077,1999-12-03 19:24:37,1999,12
1,1,25,1.0,944250228,1999-12-03 19:43:48,1999,12
2,1,29,2.0,943230976,1999-11-22 00:36:16,1999,11
3,1,30,5.0,944249077,1999-12-03 19:24:37,1999,12
4,1,32,5.0,943228858,1999-11-22 00:00:58,1999,11


In [70]:
ratings_df['time_dt']=pd.to_datetime(ratings_df['timestamp'],unit='s')

In [72]:
ratings_df['rating_year']=ratings_df['time_dt'].dt.year

In [73]:
ratings_df['rating_month']=ratings_df['time_dt'].dt.month

In [77]:
movie_rate = pd.merge(movies_df,ratings_df,how='left',on='movieId')

movie_rate.head()

,movieId,title,genres,year,(no genres listed),Action,Adventure,Animation,Children,Comedy,...,Sci-Fi,Thriller,War,Western,userId,rating,timestamp,time_dt,rating_year,rating_month
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,1995.0,0,0,1,1,1,1,...,0,0,0,0,10.0,2.5,1.169265e+09,2007-01-20 03:53:51,2007.0,1.0
1,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,1995.0,0,0,1,1,1,1,...,0,0,0,0,11.0,3.0,8.500851e+08,1996-12-08 22:44:36,1996.0,12.0
2,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,1995.0,0,0,1,1,1,1,...,0,0,0,0,17.0,4.0,1.027306e+09,2002-07-22 02:42:31,2002.0,7.0
3,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,1995.0,0,0,1,1,1,1,...,0,0,0,0,19.0,3.0,9.747045e+08,2000-11-20 07:14:48,2000.0,11.0
4,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,1995.0,0,0,1,1,1,1,...,0,0,0,0,20.0,5.0,1.553184e+09,2019-03-21 16:03:50,2019.0,3.0


In [84]:
movies_df=movies_df.loc[:, ~movies_df.columns.duplicated() ]

In [98]:
movie_rate = ratings_df.merge(movies_df, on='movieId', how='left')

In [99]:
print(f"userId Nulls: {movie_rate['userId'].isnull().sum()}")

userId Nulls: 0


In [100]:
movie_rate = movie_rate.dropna(subset=['userId'
                                        ,'rating'])

In [101]:
movie_rate['userId'] = movie_rate['userId'].astype(int)

In [102]:
#Get 10% of sample data from 3.2M population 
df_sample = movie_rate.sample(frac=0.1,random_state=42)

df_sample.head()

,userId,movieId,rating,timestamp,time_dt,rating_year,rating_month,title,genres,year,...,Film-Noir,Horror,IMAX,Musical,Mystery,Romance,Sci-Fi,Thriller,War,Western
10685861,66954,781,5.0,850944577,1996-12-18 21:29:37,1996,12,Stealing Beauty (1996),Drama,1996.0,...,0,0,0,0,0,0,0,0,0,0
1552723,9877,574,4.0,945495614,1999-12-18 05:40:14,1999,12,Spanking the Monkey (1994),Comedy|Drama,1994.0,...,0,0,0,0,0,0,0,0,0,0
6145184,38348,1088,2.0,999974867,2001-09-08 18:47:47,2001,9,Dirty Dancing (1987),Drama|Musical|Romance,1987.0,...,0,0,0,1,0,1,0,0,0,0
16268584,101952,2706,1.0,1203077565,2008-02-15 12:12:45,2008,2,American Pie (1999),Comedy|Romance,1999.0,...,0,0,0,0,0,1,0,0,0,0
22418634,140400,275079,3.5,1653782463,2022-05-29 00:01:03,2022,5,Chip 'n Dale: Rescue Rangers (2022),Adventure|Animation|Children|Comedy|Fantasy|My...,2022.0,...,0,0,0,0,1,0,0,0,0,0


In [103]:
df_sample.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3200020 entries, 10685861 to 31798448
Data columns (total 30 columns):
 #   Column              Dtype         
---  ------              -----         
 0   userId              int64         
 1   movieId             int64         
 2   rating              float64       
 3   timestamp           int64         
 4   time_dt             datetime64[ns]
 5   rating_year         int32         
 6   rating_month        int32         
 7   title               object        
 8   genres              object        
 9   year                float64       
 10  (no genres listed)  int64         
 11  Action              int64         
 12  Adventure           int64         
 13  Animation           int64         
 14  Children            int64         
 15  Comedy              int64         
 16  Crime               int64         
 17  Documentary         int64         
 18  Drama               int64         
 19  Fantasy             int64         
 20 

In [104]:
df_sample['movie_age']=df_sample['rating_year']-df_sample['year']

print(df_sample['movie_age'].describe())

count    3.196636e+06
mean     1.429662e+01
std      1.462934e+01
min     -4.000000e+00
25%      4.000000e+00
50%      1.000000e+01
75%      1.900000e+01
max      1.470000e+02
Name: movie_age, dtype: float64


In [111]:
print(f'Negative movie age rows: {(df_sample['movie_age']<0).sum()}')

Negative movie age rows: 95


### Phase 2- EDA

In [113]:
df_sample['rating'].value_counts().sort_index()

rating
0.5     52700
1.0     94681
1.5     53137
2.0    201939
2.5    169044
3.0    604716
3.5    430453
4.0    836484
4.5    297352
5.0    459514
Name: count, dtype: int64

In [116]:
ratings_per_user = df_sample.groupby('userId')['rating'].count()
ratings_per_user.describe()

count    197270.000000
mean         16.221524
std          28.686731
min           1.000000
25%           4.000000
50%           8.000000
75%          17.000000
max        3381.000000
Name: rating, dtype: float64

In [117]:
ratings_per_movie = df_sample.groupby('movieId')['rating'].count()
ratings_per_movie.describe()

count    42809.000000
mean        74.751104
std        360.557722
min          1.000000
25%          1.000000
50%          3.000000
75%         14.000000
max      10368.000000
Name: rating, dtype: float64


Ratings per movie → 42,809 unique movies
median: 3 ratings per movie
75%:    14
max:    10,368
Very sparse — most movies barely rated.

Ratings per user → 197,270 unique users
median: 8 ratings per user
75%:    17
max:    3,381
More users than movies, but still sparse behaviour.

In [122]:
active_movies = df_sample.groupby('movieId')['rating'].count()
active_users = df_sample.groupby('userId')['rating'].count()

userId
1         10
2          8
3         18
4          2
5          1
          ..
200943    71
200944    30
200945    10
200947     5
200948    21
Name: rating, Length: 197270, dtype: int64

In [126]:
#Filter and keep only movies that have atleast 5 ratings and keep only users that have rated atleast 10 movies

valid_movies = active_movies[active_movies >= 10].index  #extracts just movieIds
valid_users = active_users[active_users >= 5].index #extracts userIds


In [127]:
df_model = df_sample[
    df_sample['movieId'].isin(valid_movies) &
    df_sample['userId'].isin(valid_users)
    ]

In [129]:
print(f'Before: {len(df_sample):,} rows, {df_sample['movieId'].nunique():,} movies, {df_sample['userId'].nunique():,} users')
print(f'After: {len(df_model):,} rows, {df_model['movieId'].nunique():,} movies, {df_model['userId'].nunique():,} users')

Before: 3,200,020 rows, 42,809 movies, 197,270 users
After: 2,964,172 rows, 12,526 movies, 134,940 users


In [133]:
df_model.head()

,userId,movieId,rating,timestamp,time_dt,rating_year,rating_month,title,genres,year,...,Horror,IMAX,Musical,Mystery,Romance,Sci-Fi,Thriller,War,Western,movie_age
10685861,66954,781,5.0,850944577,1996-12-18 21:29:37,1996,12,Stealing Beauty (1996),Drama,1996.0,...,0,0,0,0,0,0,0,0,0,0.0
1552723,9877,574,4.0,945495614,1999-12-18 05:40:14,1999,12,Spanking the Monkey (1994),Comedy|Drama,1994.0,...,0,0,0,0,0,0,0,0,0,5.0
6145184,38348,1088,2.0,999974867,2001-09-08 18:47:47,2001,9,Dirty Dancing (1987),Drama|Musical|Romance,1987.0,...,0,0,1,0,1,0,0,0,0,14.0
16268584,101952,2706,1.0,1203077565,2008-02-15 12:12:45,2008,2,American Pie (1999),Comedy|Romance,1999.0,...,0,0,0,0,1,0,0,0,0,9.0
22418634,140400,275079,3.5,1653782463,2022-05-29 00:01:03,2022,5,Chip 'n Dale: Rescue Rangers (2022),Adventure|Animation|Children|Comedy|Fantasy|My...,2022.0,...,0,0,0,1,0,0,0,0,0,0.0


### Feature Engineering on df_model

In [132]:
#movie stats
movie_stats = df_model.groupby('movieId')['rating'].agg(avg_rating_per_movies='mean',rating_count_per_movie='count').reset_index()
movie_stats

,movieId,avg_rating_per_movies,rating_count_per_movie
0,1,3.895124,6050
1,2,3.245991,2744
2,3,3.131718,1135
3,4,2.787698,252
4,5,3.029439,1070
...,...,...,...
12521,288269,3.275000,20
12522,288311,3.450000,10
12523,288513,3.660920,87
12524,288861,3.166667,12


In [135]:
#user stats
user_stats = df_model.groupby('userId')['rating'].agg(avg_rating_per_user='mean',rating_count_per_user='count').reset_index()
user_stats

,userId,avg_rating_per_user,rating_count_per_user
0,1,4.000000,10
1,2,4.625000,8
2,3,3.833333,18
3,9,4.300000,5
4,10,2.730769,78
...,...,...,...
134935,200943,3.028986,69
134936,200944,3.883333,30
134937,200945,3.250000,10
134938,200947,3.900000,5


In [136]:
#merge to df_model

df_model = df_model.merge(movie_stats, on='movieId', how='left')
df_model = df_model.merge(user_stats, on='userId', how='left')

In [137]:
df_model.head()

,userId,movieId,rating,timestamp,time_dt,rating_year,rating_month,title,genres,year,...,Romance,Sci-Fi,Thriller,War,Western,movie_age,avg_rating_per_movies,rating_count_per_movie,avg_rating_per_user,rating_count_per_user
0,66954,781,5.0,850944577,1996-12-18 21:29:37,1996,12,Stealing Beauty (1996),Drama,1996.0,...,0,0,0,0,0,0.0,3.433673,196,4.285714,14
1,9877,574,4.0,945495614,1999-12-18 05:40:14,1999,12,Spanking the Monkey (1994),Comedy|Drama,1994.0,...,0,0,0,0,0,5.0,3.316794,131,3.542169,83
2,38348,1088,2.0,999974867,2001-09-08 18:47:47,2001,9,Dirty Dancing (1987),Drama|Musical|Romance,1987.0,...,1,0,0,0,0,14.0,3.246453,1339,2.825000,40
3,101952,2706,1.0,1203077565,2008-02-15 12:12:45,2008,2,American Pie (1999),Comedy|Romance,1999.0,...,1,0,0,0,0,9.0,3.274545,2750,3.220000,25
4,140400,275079,3.5,1653782463,2022-05-29 00:01:03,2022,5,Chip 'n Dale: Rescue Rangers (2022),Adventure|Animation|Children|Comedy|Fantasy|My...,2022.0,...,0,0,0,0,0,0.0,3.050000,20,3.566901,142


The problem — power law distributions
Your ratings per user look like this:
Most users:  8 ratings
Some users:  100 ratings  
Few users:   3,381 ratings
If you plot it, it's a hockey stick — a tiny number of users have extreme values. This is called a power law or long tail.

Why this hurts ML models
Most ML algorithms assume features are roughly on a similar scale. When one user has 8 ratings and another has 3,381, that gap of 3,373 dominates the model. The algorithm pays huge attention to that difference and less to everything else. This causes:

Poor generalisation
Feature importance being distorted
Coefficients being unstable


What log transform does
np.log1p(x) = log(x + 1) — the +1 handles zeros safely.
It compresses large values while preserving the order:
Raw count  →  log1p
1          →  0.69
8          →  2.20
17         →  2.94
100        →  4.62
1,000      →  6.91
3,381      →  8.13
The gap between 1 and 3,381 goes from 3,380 → 7.44. The order is preserved — a user with more ratings still has a higher value — but the extreme values no longer dominate.

Why +1 specifically?
Because log(0) is undefined (negative infinity). Adding 1 ensures the minimum input is 1, giving log(1) = 0 cleanly. That's why it's log1p not just log.

Rule of thumb: whenever you see count data or monetary values with a long tail — ratings, revenue, page views, salaries — log transform is almost always the right move before feeding into a model.
Ready to continue with the output?

In [139]:
#log transform 
df_model['log_movie_count']=np.log1p(df_model['rating_count_per_movie'])
df_model['log_user_count']=np.log1p(df_model['rating_count_per_user'])


In [148]:
df_model.head(2)

,userId,movieId,rating,timestamp,time_dt,rating_year,rating_month,title,genres,year,...,War,Western,movie_age,avg_rating_per_movies,rating_count_per_movie,avg_rating_per_user,rating_count_per_user,log_movie_count,log_user_count,liked
0,66954,781,5.0,850944577,1996-12-18 21:29:37,1996,12,Stealing Beauty (1996),Drama,1996.0,...,0,0,0.0,3.433673,196,4.285714,14,5.283204,2.708050,1
1,9877,574,4.0,945495614,1999-12-18 05:40:14,1999,12,Spanking the Monkey (1994),Comedy|Drama,1994.0,...,0,0,5.0,3.316794,131,3.542169,83,4.882802,4.430817,1


In [143]:
#movie_age
df_model['movie_age']=df_model['rating_year']-df_model['year']

In [147]:
#classification target
df_model['liked']= (df_model['rating']>3).astype(int)

In [152]:
print(df_model[['avg_rating_per_movies','rating_count_per_movie'
                ,'avg_rating_per_user','rating_count_per_user'
                ,'movie_age','liked']].describe())
print(f"\nLiked %: {df_model['liked'].mean()*100:.1f}%")

       avg_rating_per_movies  rating_count_per_movie  avg_rating_per_user  \
count           2.964172e+06            2.964172e+06         2.964172e+06   
mean            3.539721e+00            1.658601e+03         3.539721e+00   
std             4.523060e-01            1.763813e+03         5.137886e-01   
min             9.500000e-01            8.000000e+00         5.000000e-01   
25%             3.281343e+00            3.540000e+02         3.223529e+00   
50%             3.617521e+00            1.056000e+03         3.551136e+00   
75%             3.874197e+00            2.367000e+03         3.875000e+00   
max             4.600000e+00            8.732000e+03         5.000000e+00   

       rating_count_per_user     movie_age         liked  
count           2.964172e+06  2.961175e+06  2.964172e+06  
mean            6.047203e+01  1.427647e+01  6.335364e-01  
std             7.299553e+01  1.435457e+01  4.818382e-01  
min             1.000000e+00 -4.000000e+00  0.000000e+00  
25%        

### Analysis
avg_rating_per_movie
mean: 3.54,  std: 0.45,  range: 0.95 – 4.60
Tight distribution — most movies cluster around 3.5. No movie averages above 4.6 with 10+ ratings. Makes sense — extreme averages wash out with enough ratings.

rating_count_per_movie
mean: 1,658,  median: 1,056,  max: 8,732
Even after filtering to 10+ ratings, popular movies dominate heavily. The top 25% of movies have 2,367+ ratings each — these blockbusters will have very reliable average ratings.

avg_rating_per_user
mean: 3.54,  std: 0.51,  range: 0.5 – 5.0
Some users are harsh raters (avg 0.5!) and some are generous (avg 5.0). This user bias is one of the strongest signals in your model — knowing whether someone is a harsh or generous rater tells you a lot about their next rating.

rating_count_per_user
mean: 60,  median: 36,  max: 1,019
After filtering, users are much more active — median 36 ratings. Good enough history to build user features from.

movie_age
mean: 14 years,  median: 10 years,  max: 127 years!
Negative values (-4) mean rated before release — pre-release screenings or data entry errors. 127 years old means someone rated a film from ~1896. Flag these as outliers.

liked: 63.4%
This is critical for your classifier. Your classes are imbalanced — 63.4% liked vs 36.6% not liked. This means:

A dumb model that always predicts "liked" gets 63.4% accuracy for free
Accuracy alone is a useless metric — you must use F1, precision, recall, or AUC-ROC
You'll need to handle this imbalance when training

### Hypothesis Testing

<b>The hypothesis Test 1:</b>


1990 is a natural dividing line — pre-1990 films are classic/vintage, post-1990 are modern. The hypothesis is whether nostalgia or classic status inflates ratings for older films. Could have used 1980 or 2000 — the exact cutoff is a judgement call you should justify in your write-up.

H₀: Old movies (pre-1990) and new movies (post-1990) are rated the same <br>
H₁: There is a significant difference in ratings between old and new movies

In [153]:
old = df_model[df_model['year']<1990]['rating']
new = df_model[df_model['year']>=1990]['rating']

<b>Run Welch's t-test on the two groups.</b>

Returns:

> t1 — how many standard errors apart the two means are <br>
> p1 — probability of seeing this difference by chance if H₀ were true

t1: t-statistic:"How many standard errors apart are the two group means?" <br>
t = (mean_old - mean_new) / standard_error <br>

t = 0          → means are identical, no difference at all<br>
t = 1 or -1    → means are 1 standard error apart, weak signal<br>
t = 2 or -2    → getting significant (~p=0.05)<br>
t = 3+         → strong signal, almost certainly significant<br>
t = 50+        → what you'll likely see with 2.9M rows<br>

t  →  tells you direction + how strong the signal is relative to noise<br>
p  →  tells you if it's statistically significant<br>
mean difference  →  tells you if it's practically meaningful<br>

ttest_ind = t-test for two independent groups. Here's what it does under the hood:
pythont_stat, p_value = stats.ttest_ind(group_a, group_b, equal_var=False)
Step 1 — Calculate means and variances of both groups:
mean_a, var_a, n_a = old.mean(), old.var(), len(old)
mean_b, var_b, n_b = new.mean(), new.var(), len(new)
Step 2 — Calculate standard error (Welch's formula):
SE = sqrt(var_a/n_a + var_b/n_b)
Step 3 — Calculate t-statistic:
t = (mean_a - mean_b) / SE
This is just "how many standard errors apart are the two means?"
Step 4 — Calculate degrees of freedom (Welch-Satterthwaite equation):
df = (var_a/n_a + var_b/n_b)²  /  ((var_a/n_a)²/(n_a-1) + (var_b/n_b)²/(n_b-1))
This adjusts df downward when variances are unequal — making the test more conservative.
Step 5 — Get p-value from t-distribution:
p = 2 * (1 - CDF(|t|, df))   ← two-tailed

In [158]:
# perform t test of 2 independent groups old and new [not paired, seperate groups]
#equal_var=True   →  Student's t-test (assumes same variance in both groups)
#equal_var=False  →  Welch's t-test (no assumption about variance) ← safer

t1, p1 = stats.ttest_ind(old,new, equal_var=False)

Always print the means first — the t-test tells you if there's a difference, the means tell you which direction and how big.

In [168]:
print(f'Old mean: {old.mean():.3f}, New mean: {new.mean():.3f}')
print(f"t={t1:.4f}, p={p1:.6f}")
print("→ Reject H0" if p1 < 0.05 else "→ Fail to reject H0")
print()

Old mean: 3.697, New mean: 3.492
t=146.2598, p=0.000000
→ Reject H0



<b>The hypothesis Test 2:</b>


100+ = power users who rate systematically and critically. ≤10 = casual users who only bother rating movies they feel strongly about (usually loved or hated). The hypothesis is whether these groups rate differently. Note we deliberately skipped the middle (11–100) to get a clean contrast between extremes.

H₀: > 100 <br>
Ha: <= 10 ratings

In [164]:
# Test 2 — Prolific vs casual users
prolific = df_model[df_model['rating_count_per_user'] > 100]['rating']
casual = df_model[df_model['rating_count_per_user'] <= 10]['rating']
t2, p2 = stats.ttest_ind(prolific, casual, equal_var=False)
print(f"Test 2 — Prolific vs casual raters")
print(f"Prolific mean: {prolific.mean():.3f}, Casual mean: {casual.mean():.3f}")
print(f"t={t2:.4f}, p={p2:.6f}")
print("→ Reject H0" if p2 < 0.05 else "→ Fail to reject H0")

print()

Test 2 — Prolific vs casual raters
Prolific mean: 3.250, Casual mean: 3.765
t=-235.4822, p=0.000000
→ Reject H0



<b>Test 3 — Action vs Romance </b><br>
Ho: pythonaction  <br>
Ha: romance <br>
These genres attract very different audiences with different rating behaviours. Action fans tend to rate more generously, romance watchers more critically — or vice versa. The genre dummy columns make this trivial to split.

In [165]:
# Test 3 — Action vs Romance ratings
action = df_model[df_model['Action'] == 1]['rating']
romance = df_model[df_model['Romance'] == 1]['rating']
t3, p3 = stats.ttest_ind(action, romance, equal_var=False)
print(f"Test 3 — Action vs Romance")
print(f"Action mean: {action.mean():.3f}, Romance mean: {romance.mean():.3f}")
print(f"t={t3:.4f}, p={p3:.6f}")
print("→ Reject H0" if p3 < 0.05 else "→ Fail to reject H0")

print()



Test 3 — Action vs Romance
Action mean: 3.468, Romance mean: 3.540
t=-39.2399, p=0.000000
→ Reject H0



<b>Test 4 — avg_rating >= 4.0 vs < 3.0</b><br>
Ho: high_movie <br>
Ha: low_movie  <br>
This tests whether movie quality (as measured by crowd average) actually predicts individual ratings — which it should, but by how much? Again skipping the middle (3.0–4.0) to get a clean signal.

In [167]:
# Test 4 — High vs low rated movies affect on user rating
high_movie = df_model[df_model['avg_rating_per_movies'] >= 4.0]['rating']
low_movie  = df_model[df_model['avg_rating_per_movies'] < 3.0]['rating']
t4, p4 = stats.ttest_ind(high_movie, low_movie, equal_var=False)
print(f"Test 4 — High quality vs low quality movies")
print(f"High mean: {high_movie.mean():.3f}, Low mean: {low_movie.mean():.3f}")
print(f"t={t4:.4f}, p={p4:.6f}")
print("→ Reject H0" if p4 < 0.05 else "→ Fail to reject H0")

Test 4 — High quality vs low quality movies
High mean: 4.113, Low mean: 2.671
t=633.8972, p=0.000000
→ Reject H0


Test 1 — Movie age effect<br>
* Old (<1990): 3.697
* New (≥1990): 3.492
* diff: +0.205
* t=146.26
Statistically significant AND practically meaningful. Old movies are rated 0.2 stars higher on average. This is real — classic films have survivorship bias (only the good ones get watched and rated decades later) plus nostalgia effect. This feature will be useful in your model.

Test 2 — Prolific vs casual raters<br>
* Prolific (>100): 3.250   
* Casual (≤10): 3.765   
* diff: -0.515   
* t=-235.48
Strongest practical effect of all 4 tests. Prolific users rate 0.515 stars lower than casual users. Why? Casual users tend to only rate movies they loved — selection bias upward. Power users rate everything systematically and are harsher critics. Critical insight — avg_rating_per_user and rating_count_per_user will be among your most important features.

Test 3 — Action vs Romance<br>
* Action: 3.468
* Romance: 3.540
* diff: -0.072
* t=-39.24
Statistically significant but practically weak. Only 0.072 stars difference. With 2.9M rows, the t-test detects this tiny difference but it's not meaningful for modelling. Genre alone is a weak predictor — genre dummies will have low feature importance. Flag this in your write-up as a case of statistical vs practical significance.

Test 4 — High vs low quality movies<br>
* High (≥4.0): 4.113
* Low (<3.0): 2.671
* diff: +1.442
* t=633.90
Enormous effect — largest t-statistic of all. 1.44 stars difference is massive. Makes intuitive sense — if a movie's crowd average is 4+, individual ratings naturally cluster there too. avg_rating_per_movie will almost certainly be your single most important feature in the regression model.



>All tests reject H₀ at p<0.05, but this is expected with n=2.9M — any real difference will be detected. Effect size (mean difference) is the more honest metric at this scale. Tests 1, 2, and 4 show practically significant effects that justify including these as model features. Test 3 suggests genre alone is a weak signal.


### Phase4 - Modeling

In [170]:
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import (mean_squared_error, r2_score, 
                             classification_report, roc_auc_score,
                             confusion_matrix)
from sklearn.preprocessing import StandardScaler

In [171]:
# define features
genre_cols = ['Action','Adventure','Animation','Children','Comedy','Crime',
              'Documentary','Drama','Fantasy','Film-Noir','Horror','IMAX',
              'Musical','Mystery','Romance','Sci-Fi','Thriller','War','Western']

In [174]:
feature_cols = [
    'avg_rating_per_movies',   # movie quality signal
    'rating_count_per_movie', # movie popularity
    'avg_rating_per_user',    # user bias
    'rating_count_per_user',  # user activity
    'log_movie_count',        # log-transformed popularity
    'log_user_count',         # log-transformed activity
    'movie_age',              # how old is the movie
] + genre_cols


In [175]:
# drop rows with any nulls in features
df_clean = df_model[feature_cols + ['rating', 'liked']].dropna()
print(f"Modelling dataset: {df_clean.shape}")
print(f"Nulls: {df_clean.isnull().sum().sum()}")

Modelling dataset: (2961175, 28)
Nulls: 0


In [176]:
# features and targets
X = df_clean[feature_cols]
y_reg = df_clean['rating']      # regression target
y_clf = df_clean['liked']       # classification target

# train/test split — stratify on liked to preserve 63/37 class balance
X_train, X_test, y_train_reg, y_test_reg, y_train_clf, y_test_clf = train_test_split(
    X, y_reg, y_clf, test_size=0.2, random_state=42, stratify=y_clf
)

print(f"Train: {X_train.shape[0]:,} rows")
print(f"Test:  {X_test.shape[0]:,} rows")
print(f"Train liked %: {y_train_clf.mean()*100:.1f}%")
print(f"Test liked %:  {y_test_clf.mean()*100:.1f}%")

# scale features — important for logistic regression
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled  = scaler.transform(X_test)

Train: 2,368,940 rows
Test:  592,235 rows
Train liked %: 63.3%
Test liked %:  63.3%


In [177]:
# ── REGRESSION ──────────────────────────────────────────
reg = LinearRegression()
reg.fit(X_train, y_train_reg)
y_pred_reg = reg.predict(X_test)

rmse = np.sqrt(mean_squared_error(y_test_reg, y_pred_reg))
r2   = r2_score(y_test_reg, y_pred_reg)

print("=" * 40)
print("REGRESSION — Predict rating (1–5)")
print("=" * 40)
print(f"RMSE: {rmse:.4f}")
print(f"R²:   {r2:.4f}")
print(f"Baseline RMSE (always predict mean): {np.sqrt(mean_squared_error(y_test_reg, [y_train_reg.mean()]*len(y_test_reg))):.4f}")

# feature importance
coef_df = pd.DataFrame({
    'feature': feature_cols,
    'coefficient': reg.coef_
}).sort_values('coefficient', ascending=False)
print("\nTop 5 features:")
print(coef_df.head())
print("\nBottom 5 features:")
print(coef_df.tail())

REGRESSION — Predict rating (1–5)
RMSE: 0.8394
R²:   0.3621
Baseline RMSE (always predict mean): 1.0510

Top 5 features:
                  feature  coefficient
2     avg_rating_per_user     0.901880
0   avg_rating_per_movies     0.858700
5          log_user_count     0.044000
19                Musical     0.022892
17                 Horror     0.017822

Bottom 5 features:
            feature  coefficient
7            Action    -0.006248
8         Adventure    -0.006348
22           Sci-Fi    -0.011537
4   log_movie_count    -0.015246
18             IMAX    -0.049176


In [178]:
# ── CLASSIFICATION ──────────────────────────────────────
clf = LogisticRegression(max_iter=1000, random_state=42)
clf.fit(X_train_scaled, y_train_clf)
y_pred_clf  = clf.predict(X_test_scaled)
y_pred_prob = clf.predict_proba(X_test_scaled)[:, 1]

print("=" * 40)
print("CLASSIFICATION — Predict liked (0/1)")
print("=" * 40)
print(classification_report(y_test_clf, y_pred_clf, 
                             target_names=['Not liked', 'Liked']))
print(f"AUC-ROC: {roc_auc_score(y_test_clf, y_pred_prob):.4f}")
print(f"Baseline accuracy (always predict liked): {y_test_clf.mean()*100:.1f}%")
print("\nConfusion Matrix:")
print(confusion_matrix(y_test_clf, y_pred_clf))

CLASSIFICATION — Predict liked (0/1)
              precision    recall  f1-score   support

   Not liked       0.70      0.55      0.62    217070
       Liked       0.77      0.87      0.81    375165

    accuracy                           0.75    592235
   macro avg       0.74      0.71      0.72    592235
weighted avg       0.74      0.75      0.74    592235

AUC-ROC: 0.8057
Baseline accuracy (always predict liked): 63.3%

Confusion Matrix:
[[119480  97590]
 [ 50350 324815]]


### Conclusion
* Liked recall = 87% — model catches 87% of movies the user will like. Great for recommendations — you don't want to miss movies someone would love.
* Not liked recall = 55% — model misses 45% of movies the user won't like. This is your weakness — it over-recommends.
* This is expected given class imbalance (63% liked) — model is biased toward predicting liked.
* RMSE improved 20% over baseline — your model predicts within ±0.84 stars on average vs the dumb baseline of always predicting the mean (±1.05 stars). Not bad for a linear model with no collaborative filtering.
* R² of 0.36 means your features explain 36% of the variance in ratings. The remaining 64% is driven by things you don't have — personal taste, mood, context. This is actually expected and honest for rating prediction.


#### Strengths:

* 20% RMSE improvement over baseline
* AUC-ROC 0.80 with a simple linear baseline model
* Confirmed hypothesis testing findings — user bias and movie quality dominate
* Stratified split preserved class balance perfectly

In [179]:
#Future scope to overcome limitations
# 1. Try Random Forest or Gradient Boosting — will capture non-linear interactions
from sklearn.ensemble import GradientBoostingClassifier

# 2. Handle class imbalance explicitly
clf = LogisticRegression(class_weight='balanced', max_iter=1000)

# 3. The real ceiling-breaker — collaborative filtering (matrix factorisation)
# Linear models can't capture "users like Alice tend to agree with users like Bob"
# SVD or ALS would be the proper next step for a recommender system

# 4. Add tag-based NLP features (TF-IDF on tags_combined)

### Merge into one working dataframe